Description: build fixes for GCC 11
Origin: upstream, https://code.qt.io/cgit/qt/qtdeclarative.git/commit/?id=eb6525f126f680f9
Last-Update: 2021-01-26

--- a/src/qml/compiler/qv4bytecodegenerator_p.h
+++ b/src/qml/compiler/qv4bytecodegenerator_p.h
@@ -186,13 +186,13 @@ QT_WARNING_POP
 
     Q_REQUIRED_RESULT Jump jumpNotUndefined()
     {
-        Instruction::JumpNotUndefined data;
+        Instruction::JumpNotUndefined data{};
         return addJumpInstruction(data);
     }
 
     Q_REQUIRED_RESULT Jump jumpNoException()
     {
-        Instruction::JumpNoException data;
+        Instruction::JumpNoException data{};
         return addJumpInstruction(data);
     }
 
