Description: fix arrays iteration to work around QTBUG-61579
Author: Dmitry Shachnev <mitya57@ubuntu.com>
Bug: https://bugreports.qt.io/browse/QTBUG-61579
Forwarded: no
Last-Update: 2019-03-14

--- a/src/imports/testlib/TestCase.qml
+++ b/src/imports/testlib/TestCase.qml
@@ -1842,7 +1842,7 @@ Item {
             testList.sort()
         }
 
-        for (var index in testList) {
+        for (var index = 0; index < testList.length; ++index) {
             var prop = testList[index]
 
             if (checkNames && !(prop in testsToRun))
@@ -1863,11 +1863,11 @@ Item {
                     let checkTags = (checkNames && testsToRun[prop].length > 0)
 
                     qtest_results.initTestTable()
-                    for (var index in table) {
+                    for (var rowIndex in table) {
                         haveData = true
-                        var row = table[index]
+                        var row = table[parseInt(rowIndex)]
                         if (!row.tag)
-                            row.tag = "row " + index    // Must have something
+                            row.tag = "row " + rowIndex    // Must have something
                         if (checkTags) {
                             let tags = testsToRun[prop]
                             let tagIdx = tags.indexOf(row.tag)
